### ISSUE TITLE
why SPR loss is calculated by MSE but not cosine similarity? 

### ISSUE THREAD

**Issue by yueyang130** [mila-iqia/spr#27]
As mentioned in the title, I found SPR loss is calculated by MSE in the code, see [SPR loss](https://github.com/yueyang130/cycle-spr/blob/88db1cc2d7c03d0ce330603f5d5641c3bd9fd113/src/models.py#L310)

However, in the paper, it says the spr loss is calculated by consine similarity.

![image](https://user-images.githubusercontent.com/55612720/146885516-b3a57a96-a57c-42a4-a94c-e8969e25cf0d.png)

So can you give some explanation the inconsistency? Is the MSE leading to better/ more stable performance? Or some other reason?

Thanks!

**Comment by ankeshanand** [comment]
We use a normalized L2 loss in the code, its gradients are equivalent to cosine similarity upto a scaling factor. See https://stats.stackexchange.com/a/146279

**Comment by slerman12** [comment]
Any reason for this over cosine similarity?

**Comment by ankeshanand** [comment]
Like I said, they are essentially the same, it's just an implementation artifact.

**Comment by slerman12** [comment]
Fair enough. I'm not sure the stack overflow link confirms that though. In any case, should this line be detached from the gradient graph?

`f_x2 = F.normalize(f_x2s.float(), p=2., dim=-1, eps=1e-3)`

In Bootstrap Your Own Latent, I believe they detach the computation of the target.

**Comment by slerman12** [comment]
Just checking. I'm really not sure.

**Comment by ankeshanand** [comment]
Just to be clear, `norm_l2_loss = 2 + 2*(-cos_sim)`, so their gradients will be proportional.

The targets are detached here: https://github.com/mila-iqia/spr/blob/release/src/models.py#L298

**Comment by slerman12** [comment]
Okay, just making sure it's not a rounding thing too. Don't know if using Pytorch's cosine similarity somehow risks a rounding error or something. Thanks for clarifying.

**Comment by slerman12** [comment]
I'm a bit confused, what is the difference between this L2 normalization and the "re-normalize" operation? I guess this one is applied to the predictor and projector heads, while the other is applied to... the CNN features of the encoder and transition model? Is re-normalize also used during training in other places or just for supervising the dynamics loss?

**Comment by yueyang130** [comment]
@ankeshanand Thanks very much! I understand.

**Comment by slerman12** [comment]
Sorry, just to get back to my question, is there a min-max [0, 1] normalization applied to the CNN features? There's a re-normalize operation here that seems to be applied to the CNN encoder?

**Comment by MaxASchwarzer** [comment]
Yeah, it's applied there and also to the output of the transition model, to stop its activations from diverging.

**Comment by slerman12** [comment]
Thanks Max. I see. This is both for action selection and SPR or just SPR?

**Comment by MaxASchwarzer** [comment]
We do it everywhere
